Perturbed optimizers notebook and narrative documentation. (#371)
google/jaxopt: docs/notebooks/perturbed_optimizers/perturbed_optimizers.md

---
jupytext:
  formats: ipynb,md:myst
  text_representation:
    extension: .md
    format_name: myst
    format_version: 0.13
    jupytext_version: 1.14.4
kernelspec:
  display_name: base
  language: python
  name: python3
---

Copyright 2022 Google LLC

Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at

 https://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.

+++

# Perturbed optimizers

+++

[![Open in Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/google/jaxopt/blob/main/docs/notebooks/perturbed_optimizers/perturbed_optimizers.ipynb)

+++

> We propose in this work a universal method to transform any optimizer $y^*$ in a differentiable approximation $y_\varepsilon^*$, using pertutbations following the method of [Berthet et al. (2020)](https://arxiv.org/abs/2002.08676). We provide a JAX implementation, illustrated here on some examples.

Reviewer: I would also change "We provide a JAX implementation, illustrated here on some examples." by "JAXopt provides an implementation that we illustrate here on some examples"

> We propose in this work a universal method to transform any optimizer $y^*$ in a differentiable approximation $y_\varepsilon^*$, using pertutbations following the method of [Berthet et al. (2020)](https://arxiv.org/abs/2002.08676). We provide a JAX implementation, illustrated here on some examples.

Author: done

> We propose in this work a universal method to transform any optimizer $y^*$ in a differentiable approximation $y_\varepsilon^*$, using pertutbations following the method of [Berthet et al. (2020)](https://arxiv.org/abs/2002.08676). We provide a JAX implementation, illustrated here on some examples.

Author: done

Concretely, for an optimizer function $y^*$ defined by

> $$y^*(\theta) = \mathop{\mathrm{arg\,max}}_{y\in \mathcal{C}} \langle y, \theta \rangle$$

Reviewer 2: ```suggestion
$$y^*(\theta) = \mathop{\mathrm{arg\,max}}_{y\in \mathcal{C}} \langle y, \theta \rangle,$$
```

Author: done

Author: done

We consider, for a random $Z$ drawn from a distribution with continuous positive distribution $\mu$

$$y_\varepsilon^*(\theta) = \mathbf{E}[\mathop{\mathrm{arg\,max}}_{y\in \mathcal{C}} \langle y, \theta + \varepsilon Z \rangle]$$

```{code-cell} ipython3
:id: 9WIWwRdSU51j

%%capture
%pip install jaxopt
```

```{code-cell} ipython3
:id: S6tLyyy9VCEw

import jax
import jax.numpy as jnp
import jaxopt
import time

from jaxopt import perturbations
```

+++ {"id": "EmYn_jNUFfw2"}

# Argmax one-hot

+++

> We start from an original optimizer, an `argmax_one_hot` function, corresponding to $y^*$ for $\mathcal{C}$ being the unit simplex. We run it on an example input `values`

Reviewer: What does "original" mean in this context? It is not clear to me

Reviewer: I think it would also be good to explain with words what the `argmax_one_hot` does. Something along the lines of:

"We consider the `argmax_one_hot` function, which transforms a real-valued vector into a binary vector with one in the coefficient with largest magnitude and zero elsewhere"

Author: done

+++ {"id": "84N-wAJ8GDK2"}

## One-hot function

```{code-cell} ipython3
:id: kMZnzhX4FjGj

def argmax_one_hot(x, axis=-1):
  return jax.nn.one_hot(jnp.argmax(x, axis=axis), x.shape[axis])
```

```{code-cell} ipython3
---
colab:
  base_uri: https://localhost:8080/
id: iynCk8734Wiz
outputId: 31e902b6-2dfa-4340-b11e-955b422bd313
---
values = jnp.array([-0.6, 1.9, -0.2, 1.1, -1.0])

one_hot_vec = argmax_one_hot(values)
print(one_hot_vec)
```

+++ {"id": "6rbNt-6zGb-J"}

## One-hot with pertubations

+++

> Our implementation allows to modify this function by creating a perturbed optimizer `pert_one_hot`, corresponding to $y_\varepsilon^*$ using Gumbel noise for $Z$.

Reviewer: ```suggestion
Our implementation transforms the `argmax_one_hot` function into perturbed one that we call `pert_one_hot`. In this case we use Gumbel noise for the perturbation.
```

```{code-cell} ipython3
:id: 7hQz6zuPwkpZ

> num_samples = 100000

Reviewer: ```suggestion
NUM_SAMPLES = 100_000
```
it's a convention to write global constants in ALL_CAPS

> sigma = 0.5

Reviewer: ```suggestion
SIGMA = 0.5
```

(and adapt the code after to use these new variable names)

rng = jax.random.PRNGKey(1)
pert_one_hot = perturbations.make_perturbed_argmax(argmax_fun=argmax_one_hot,
                                         num_samples=num_samples,
                                         sigma=sigma,
                                         noise=perturbations.Gumbel())
```

> In this particular case, it is equal to the usual softmax, with exponential weights. This is not always true, in general there is no closed form for $y_\varepsilon^*$

Reviewer: ```suggestion
In this particular case, the perturbed argmax is equivalent to the [softmax function](https://en.wikipedia.org/wiki/Softmax_function). This is not always true, in general there is no closed form for $y_\varepsilon^*$
```

```{code-cell} ipython3
---
colab:
  base_uri: https://localhost:8080/
id: f2gDpghJYZ33
outputId: adbe4b3c-4ec9-4f18-ea75-07619fd84cb3
---
rngs = jax.random.split(rng, 2)

rng = rngs[0]

pert_argmax = pert_one_hot(values, rng)
print(f'computation with {num_samples} samples, sigma = {sigma}')
print(f'perturbed argmax = {pert_argmax}')
jax.nn.softmax(values/sigma)
soft_max = jax.nn.softmax(values/sigma)
print(f'softmax = {soft_max}')
print(f'square norm of softmax = {jnp.linalg.norm(soft_max):.2e}')
print(f'square norm of difference = {jnp.linalg.norm(pert_argmax - soft_max):.2e}')
```

+++ {"id": "2U7rhtEAGpMV"}

## Gradients for one-hot with perturbations

+++

The perturbed optimizer $y_\varepsilon^*$ is differentiable, and its gradient can be computed with stochastic estimation automatically, using `jax.grad`.

> We create a scalar loss `loss_simplex` of the perturbed optimizer $y^*_\varepsilon$

Reviewer: perhaps worth saying that this is a squared loss between y_true and y_eps ? This was not clear to me from the text + math
